File: 4367-minimum-cost-path-with-at-most-k-turns/minimum-cost-path-with-at-most-k-turns.cpp
class Solution {
public:
    int minCost(vector<vector<int>>& grid, int k) {
        int m = grid.size();
        int n = grid[0].size();
        const int INF = 1e9;
        vector<vector<vector<vector<int>>>>dist(
            m,
            vector<vector<vector<int>>>(
                n,
                vector<vector<int>>(
                    5,
                    vector<int>(k+1, INF)
                )
            )
        );
        using state = tuple<int, int, int, int, int>;
        priority_queue<state, vector<state>, greater<state>>pq;
        dist[0][0][4][0] = grid[0][0];
        pq.push({grid[0][0], 0, 0, 4, 0});
        int dr[4] = {-1, 1, 0, 0};
        int dc[4] = {0, 0, -1, 1};
        while(!pq.empty()){
            auto [cost, r, c, prevDir, turns] = pq.top();
            pq.pop();
            if(cost != dist[r][c][prevDir][turns]){
                continue;
            }
            if(r == m-1 && c == n-1){
                return cost;
            }
            for(int dir = 0; dir<4; dir++){
                int nr = r+dr[dir];
                int nc = c+dc[dir];
                if(nr<0 || nr>=m || nc<0 || nc>=n){
                    continue;
                }
                int newTurns = turns;
                if(prevDir != 4 && prevDir != dir){
                    newTurns++;
                }
                if(newTurns > k){
                    continue;
                }
                int newCost = cost + grid[nr][nc];
                if(newCost < dist[nr][nc][dir][newTurns]){
                    dist[nr][nc][dir][newTurns] = newCost;
                    pq.push({newCost, nr, nc, dir, newTurns});
                }
            }
        }
        return -1;
    }
};

File: 4367-minimum-cost-path-with-at-most-k-turns/minimum-cost-path-with-at-most-k-turns.py
class Solution:
    def minCost(self, grid: list[list[int]], k: int) -> int:
        m = len(grid)
        n = len(grid[0])
        INF = 10**9
        dist = [
            [
                [[INF]*(k+1) for _ in range(5)]
                for _ in range(n)
            ]
            for _ in range(m)
        ]
        pq = []
        dist[0][0][4][0] = grid[0][0]
        heapq.heappush(pq, (grid[0][0],0,0,4,0))
        dr = [-1, 1, 0, 0]
        dc = [0, 0, -1, 1]
        while pq:
            cost, r, c, prev_dir, turns = heapq.heappop(pq)
            if cost != dist[r][c][prev_dir][turns]:
                continue
            if r == m-1 and c == n-1:
                return cost
            for direction in range(4):
                nr = r + dr[direction]
                nc = c + dc[direction]
                if nr < 0 or nr >=m or nc < 0 or nc >= n:
                    continue
                new_turns = turns
                if prev_dir != 4 and prev_dir != direction:
                    new_turns+=1
                if new_turns > k:
                    continue
                new_cost = cost + grid[nr][nc]
                if new_cost < dist[nr][nc][direction][new_turns]:
                    dist[nr][nc][direction][new_turns] = new_cost
                    heapq.heappush(
                        pq,
                        (new_cost, nr, nc, direction, new_turns)
                    )
        return -1